File: python/tree/leetocde/binary_tree_level_order_traversal.py

# 二叉树的层次遍历
# https://leetcode-cn.com/problems/binary-tree-level-order-traversal/
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:

    # 广度优先遍历的方式
    def levelOrder_1(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue, res = [root], []
        while len(queue) > 0:
            n, temp = len(queue), []
            for _ in range(n):
                node = queue.pop(0)
                temp.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            res.append(temp)
        return res

    # 深度优先遍历的方式
    def levelOrder_2(self, root: TreeNode) -> List[List[int]]:
        def dfs(node: TreeNode, level, res):
            if not node:
                return
            if level == len(res):
                res.append([])
            res[level].append(node.val)
            dfs(node.left, level + 1, res)
            dfs(node.right, level + 1, res)
        r = []
        dfs(root, 0, r)
        return r

File: go/tree/leetcode/binary_tree_level_order_traversal.go
package leetcode

//二叉树的层次遍历

//广度优先
func binaryTreeLevelOrder1(root *TreeNode) [][]int {
	var res [][]int
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		var lis []int
		n := len(queue)

		for i := 0; i < n; i++ {
			node := queue[0]
			queue = queue[1:]
			lis = append(lis, node.Val)

			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		res = append(res, lis)
	}

	return res
}

//深度优先
func levelOrder2(root *TreeNode) [][]int {
	var res [][]int
	levelOrderHelper(root, 0, &res)
	return res
}

func levelOrderHelper(root *TreeNode, level int, res *[][]int) {
	if root == nil {
		return
	}

	if level == len(*res) {
		*res = append(*res, []int{})
	}

	(*res)[level] = append((*res)[level], root.Val)
	levelOrderHelper(root.Left, level+1, res)
	levelOrderHelper(root.Right, level+1, res)
}
